pkg/services/pluginsintegration/pluginclient/pluginclient.go

```diff
@@ -0,0 +1,219 @@
+		ref.datasource = ds
+
+		if ds != nil {
+			ref.datasourceID = ds.ID
+			ref.datasourceUID = ds.UID
+		}
+	})
+}
+
+// DatasourceRef creates a new datasource PluginReference.
+func DatasourceRef(pluginID string, opts ...DatasourceReferenceOption) PluginReference {
+	ref := &datasourceRef{
+		pluginID: pluginID,
+	}
+
+	for _, opt := range opts {
+		opt(ref)
+	}
+
+	return ref
+}
+
+type QueryDataRequest struct {
+	Reference PluginReference
+
+	// Headers the environment/metadata information for the request.
+	//
+	// To access forwarded HTTP headers please use
+	// GetHTTPHeaders or GetHTTPHeader.
+	Headers map[string]string
+
+	// Queries the data queries for the request.
+	Queries []backend.DataQuery
+}
+
+// QueryDataHandlerFunc is an adapter to allow the use of
+// ordinary functions as QueryDataHandler. If f is a function
+// with the appropriate signature, QueryDataHandlerFunc(f) is a
+// Handler that calls f.
+type QueryDataHandlerFunc func(ctx context.Context, req *QueryDataRequest) (*backend.QueryDataResponse, error)
+
+// QueryData calls fn(ctx, req).
+func (fn QueryDataHandlerFunc) QueryData(ctx context.Context, req *QueryDataRequest) (*backend.QueryDataResponse, error) {
+	return fn(ctx, req)
+}
+
+type CheckHealthRequest struct {
+	Reference PluginReference
+
+	// Headers the environment/metadata information for the request.
+	//
+	// To access forwarded HTTP headers please use
+	// GetHTTPHeaders or GetHTTPHeader.
+	Headers map[string]string
+}
+
+type CallResourceRequest struct {
+	Reference PluginReference
+
+	// Path the forwarded HTTP path for the request.
+	Path string
+
+	// Method the forwarded HTTP method for the request.
+	Method string
+
+	// URL the forwarded HTTP URL for the request.
+	URL string
+
+	// Headers the forwarded HTTP headers for the request, if any.
+	//
+	// Recommended to use GetHTTPHeaders or GetHTTPHeader
+	// since it automatically handles canonicalization of
+	// HTTP header keys.
+	Headers map[string][]string
+
+	// Body the forwarded HTTP body for the request, if any.
+	Body []byte
+
+	// Validate if provided, will be used to validate the backend.PluginContext.
+	Validate func(ctx context.Context, pCtx backend.PluginContext) error
+}
+
+// CallResourceHandlerFunc is an adapter to allow the use of
+// ordinary functions as CallResourceHandler. If f is a function
+// with the appropriate signature, CallResourceHandlerFunc(f) is a
+// Handler that calls f.
+type CallResourceHandlerFunc func(ctx context.Context, req *CallResourceRequest, sender backend.CallResourceResponseSender) error
+
+// QueryData calls fn(ctx, req).
+func (fn CallResourceHandlerFunc) CallResource(ctx context.Context, req *CallResourceRequest, sender backend.CallResourceResponseSender) error {
+	return fn(ctx, req, sender)
+}
+
+type CollectMetricsRequest struct {
+	Reference PluginReference
+}
+
+// Client provider the client interface for interacting with plugins.
+type Client interface {
```

Comment:
Rather than having the input to this let you change the plugin... can it be *pre* scoped the the plugin you are working with?  and fail/ignore the existing plugin context?

Then we don't even need a new interface IIUC?

Reply:
You mean like a factory, e.g. `clientfactory.Get(<reference>) (error, plugin.Client)`?